walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,151 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.lang.Object;
+import java.util.TreeMap;
+import java.util.Set;
+import java.util.Map;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    //The TreeMap stores pairs of key and value. It sorts pairs in ascending order based on their key.
+    //Key is start time of event and value is event. 
+    //So Events will be sorted based on their start time when they get added to the TreeMap.   
+    TreeMap<Integer, Event> eventsList = new TreeMap<Integer, Event>();
+    
+    //Go through events attendees and request attendees 
+    //and add only the important events to TreeMap.
+    //Important events are events that contain any of the attendees in the request attendees. 
+
+    Iterator i = events.iterator();
+    Iterator iter = request.getAttendees().iterator();
+
+    while (i.hasNext()) {
+        Event currEvent = (Event)i.next();
+        //if meeting request only has one attendee
+        if (request.getAttendees().size() == 1) { 
+            Iterator iterAttendee = request.getAttendees().iterator();
+            String oneAttendee = (String)iterAttendee.next();
+            if (currEvent.getAttendees().contains(oneAttendee)) {
+                eventsList.put(currEvent.getWhen().start(), currEvent);
+            }
+        }
+        //if meeting request contains more than one attendee
+        else { 
+            while (iter.hasNext()) {
+                String currentAttendee = (String)iter.next();
+                if (currEvent.getAttendees().contains(currentAttendee)) {
+                    eventsList.put(currEvent.getWhen().start(), currEvent);
+                    break;
+                }
+            }
+        }
+    }
+    
+    //Need this in order to iterate through the TreeMap
+    Set set = eventsList.entrySet();
+    
+    //finalTimes will contain all the times the meeting can happen
+    ArrayList<TimeRange> finalTimes = new ArrayList<TimeRange>();
+   
+    //If meeting duration > 24 hours, return no options
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        finalTimes.clear();
+    } 
+
+    //If request has no attendees or there are no events already set up, return whole day
+    else if (request.getAttendees().isEmpty() || events.isEmpty() || eventsList.isEmpty()) {
+        finalTimes.add(TimeRange.WHOLE_DAY);
+    }
+
+    int limitTime = TimeRange.START_OF_DAY;
+    boolean contained = false;
+    boolean overlapped = false;
+    Event nextEvent = null;
+
+    //Iterates through sorted important events to find the meeting times
+    for (Iterator iterator = set.iterator(); iterator.hasNext();) {
```

Comment:
Can also be written with forEach loop

`for (Map.Entry entry : eventsList.entrySet()) { `